File: interpreter.py
INTEGER, OPERATOR, EOF = 'INTEGER', 'OPERATOR', 'EOF'
OPERATORS = ['+', '-', '*', '/']

class Token(object):
    def __init__(self, type, value):
        self.type = type
        self.value = value

    def __str__(self):
        return f'Token({self.type}, {self.value})'

    def __repr__(self):
        return self.__str__()


class Interpreter(object):
    def __init__(self, text):
        self.text = text
        self.pos = 0
        self.current_token = None
        self.current_char = self.text[self.pos]

    def error(self):
        raise Exception('Error parsing input')

    def get_next_token(self) -> Token:
        text = self.text

        if self.pos >= len(text) or self.current_char is None:
            return Token(EOF, None)

        if self.current_char.isdigit():
            return Token(INTEGER, self.integer())
        
        if self.current_char in OPERATORS:
            token = Token(OPERATOR, self.current_char)
            self.advance()
            return token

        #Recurse the function until whitespace is not found
        if self.current_char.isspace():
            self.advance()
            return self.get_next_token()
            
        self.error()

    def eat(self, token_type):
        if self.current_token.type == token_type:
            self.current_token = self.get_next_token()
        else:
            self.error()

    def advance(self):
        self.pos += 1
        if self.pos >= len(self.text): #End of the text
            self.current_char = None
        else:
            self.current_char = self.text[self.pos]

    def term(self):
        token = self.current_token
        self.eat(INTEGER)
        return token.value

    def integer(self):
        result = ''
        while self.current_char is not None and self.current_char.isdigit():
            result += self.current_char
            self.advance()
        return int(result)

    def expr(self):
        self.current_token = self.get_next_token()

        result = self.term()
        while self.current_token.type == OPERATOR:
            if self.current_token.value == '+':
                self.eat(OPERATOR)
                result = result + self.term()
            elif self.current_token.value == '-':
                self.eat(OPERATOR)
                result = result - self.term()
            elif self.current_token.value == '*':
                self.eat(OPERATOR)
                result = result * self.term()
            elif self.current_token.value == '/':
                self.eat(OPERATOR)
                result = result / self.term()
        
        return result


def main():
    while True:
        try:
            text = input('calc> ')
        except EOFError:
            break

        if not text:
            continue
        interpreter = Interpreter(text)
        result = interpreter.expr()
        print(result)

main()

File: interpreter.c
/*
  char -> [lexer] -> token -> [parser|interpreter] -> result
*/

#include <stdio.h>
#include <ctype.h>
#include <stdlib.h>
#include <string.h>

enum TokenTypes {_EOF, OPERATOR, INTEGER};
enum Operators {ADD, SUBTRACT, MULTIPLY, DIVIDE};

struct Token {
    enum TokenTypes type;
    int value;
};

struct Token get_next_token();

int position;
int string_len;
char current_char;
char *input_string;
struct Token current_token;

void init(char *string)
{
    position = 0;
    current_char = string[position];
    string_len = strlen(string);

    input_string = calloc(string_len, sizeof(char));
    strcpy(input_string, string);
    free(string);
}

struct Token token(enum TokenTypes token_type, int value)
{
    struct Token tk;
    tk.type = token_type;
    tk.value = value;
    return tk;
}

_Bool isoperator(char ch)
{
    char operator_syms[] = {'+', '-', '*', '/'};
    for (size_t i = 0; i < 4; i++)
    {
        if (operator_syms[i] == ch)
        {
            return 1;
        }
    }
    return 0;
}

void advance()
{
    position++;
    if (position <= string_len)
    {
        current_char = input_string[position];
    }
}

int parse_int()
{
    int result = 0;
    while (isdigit(current_char))
    {
        result *= 10;
        result += current_char - '0';
        advance();
    }
    return result;
}

struct Token get_next_token()
{
    if (position >= string_len)
    {
        return token(_EOF, 0);
    }

    if (isdigit(current_char))
    {
        return token(INTEGER, parse_int());
    }

    if (isoperator(current_char))
    {
        struct Token tk;
        switch (current_char)
        {
        case '+':
            tk = token(OPERATOR, ADD);
            break;
        case '-':
            tk = token(OPERATOR, SUBTRACT);
            break;
        case '*':
            tk = token(OPERATOR, MULTIPLY);
            break;
        case '/':
            tk = token(OPERATOR, DIVIDE);
            break;
        default:
            break;
        }
        advance();
        return tk;
    }

    advance();
    return get_next_token();
}

void eat(enum TokenTypes token_type)
{
    if (token_type == current_token.type)
    {
        current_token = get_next_token();
    }
}

int term()
{
    int val = current_token.value;
    eat(INTEGER);
    return val;
}

int parse()
{
    current_token = get_next_token();

    int result = term();
    while (current_token.type == OPERATOR)
    {
        switch (current_token.value)
        {
        case ADD:
            eat(OPERATOR);
            result = result + term();
            break;

        case SUBTRACT:
            eat(OPERATOR);
            result = result - term();
            break;

        case MULTIPLY:
            eat(OPERATOR);
            result = result * term();
            break;
        
        case DIVIDE:
            eat(OPERATOR);
            result = result / term();
            break;
        
        default:
            break;
        }
    }
    return result;
}

int main(void)
{
    while (1)
    {
        char *input = calloc(100, sizeof(char));
        printf("calc> ");
        scanf("%[^\n]%*c", input);
        init(input);
        printf("%d\n", parse());
    }
}